lib/scope-helpers.coffee

```diff
@@ -2,16 +2,39 @@ slick = require 'atom-slick'
 
 EscapeCharacterRegex = /[-!"#$%&'*+,/:;=?@|^~()<>{}[\]]/g
 
+cachedMatchesBySelector = new WeakMap
+
+getCachedMatch = (selector, scopeChain) ->
+  if cachedMatchesByScopeChain = cachedMatchesBySelector.get(selector)
+    return cachedMatchesByScopeChain[scopeChain]
+
+setCachedMatch = (selector, scopeChain, match) ->
+  unless cachedMatchesByScopeChain = cachedMatchesBySelector.get(selector)
+    cachedMatchesByScopeChain = {}
+    cachedMatchesBySelector.set(selector, cachedMatchesByScopeChain)
+  cachedMatchesByScopeChain[scopeChain] = match
+
 parseScopeChain = (scopeChain) ->
   scopeChain = scopeChain.replace EscapeCharacterRegex, (match) -> "\\#{match[0]}"
   scope for scope in slick.parse(scopeChain)[0] ? []
 
 selectorForScopeChain = (selectors, scopeChain) ->
   for selector in selectors
-    scopes = parseScopeChain(scopeChain)
-    while scopes.length > 0
-      return selector if selector.matches(scopes)
-      scopes.pop()
+    cachedMatch = getCachedMatch(selector, scopeChain)
+    if cachedMatch?
+      if cachedMatch
```

Comment:
Is V8 happy with nullable bool pattern ? undefinable bool ??
If not this migth be a deopt, since you are chasing them.